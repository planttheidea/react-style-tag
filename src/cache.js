// external dependencies
import compose from 'lodash/fp/compose';
import isUndefined from 'lodash/isUndefined';
import keys from 'lodash/keys';
import reduce from 'lodash/fp/reduce';
import uuid from 'uuid/v4';

let cache = {};

/**
 * @function createGetCacheValuesFromKeys
 *
 * @description
 * prebuilt reduce function which only requires the keys to operate
 *
 * @param {Object} [localCache=cache] the cache to retrieve values from
 * @returns {function(Array<string>): Array<string>} the reduce function to get the values based on keys
 */
export const createGetCacheValuesFromKeys = (localCache = cache) => {
  return reduce((values, key) => {
    return !localCache[key] ? values : values.concat([
      localCache[key]
    ]);
  }, []);
};

/**
 * @function getActiveValuesFromCache
 *
 * @description
 * get the ids that have active values in cache
 *
 * @param {Object} localCache the cache to get the active values from
 * @returns {Array<string>} the values in cache
 */
export const getActiveValuesFromCache = (localCache) => {
  return compose(createGetCacheValuesFromKeys(localCache), keys)(localCache);
};

/**
 * @function stylesExist
 *
 * @description
 * determine if the style passed already exists
 *
 * @param {Object} localCache the current cache of styles
 * @param {string} id the id of the style in question
 * @param {string} value the value to test for
 * @returns {boolean} does the style already exist
 */
export const stylesExist = (localCache, id, value) => {
  return !!(document.getElementById(id) || ~getActiveValuesFromCache(localCache).indexOf(value));
};

/**
 * @function createIdForTag
 *
 * @description
 * create a unique ID for the style tag
 *
 * @param {string} id the id of the style in question
 * @param {string} value the value to assign in cache
 * @param {Object} [localCache=cache] the cache to create an id for
 * @returns {null|string} the id of the item in cache
 */
export const createIdForTag = (id, value, localCache = cache) => {
  const finalId = isUndefined(id) ? uuid() : id;

  if (stylesExist(localCache, finalId, value)) {
    return null;
  }

  localCache[finalId] = value;

  return finalId;
};

/**
 * @function removeIdFromCache
 *
 * @description
 * remove the ID from cache
 *
 * @param {string} id id to remove from cache
 * @param {Object} [localCache=cache] cache to remove id from
 */
export const removeIdFromCache = (id, localCache = cache) => {
  delete localCache[id];
};

/**
 * @function setCacheId
 *
 * @description
 * update the cached value
 *
 * @param {string} id id to set in cache
 * @param {string} value value to assign to id in cache
 * @param {Object} [localCache=cache] cache to remove id from
 */
export const setCacheId = (id, value, localCache = cache) => {
  localCache[id] = value;
};
