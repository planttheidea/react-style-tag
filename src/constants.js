// external dependencies
import autoprefixer from 'autoprefixer';
import postcss from 'postcss';

/**
 * @constant {Object} REACT_STYLE_TAG_GLOBAL_PROPERTIES
 */
export const DEFAULT_REACT_STYLE_TAG_GLOBAL_PROPERTIES = {
  doNotPrefix: false,
  hasSourceMap: false,
  isMinified: false
};

/**
 * @constant {string} NO_BLOB_SUPPORT_ERROR
 * @default
 */
export const NO_BLOB_SUPPORT_ERROR = (`
To support sourcemaps for react-style-tag you need Blob support, and the browser you are using does not currently
support it. Please import the included polyfill at 'react-style-tag/blob-polyfill.js'.
`).replace(/\r?\n|\r/, '');

/**
 * @constant {string} ONLY_TEXT_ERROR
 * @default
 */
export const ONLY_TEXT_ERROR = 'The only type of child that can be used in the <Style/> tag is text.';

/**
 * @constants {Object} PREFIXER
 */
export const PREFIXER = postcss([
  autoprefixer({
    remove: false
  })
]);
