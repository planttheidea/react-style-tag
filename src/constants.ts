import { version } from 'react';

/**
 * the options to pass to cssbeautify
 */
export const BEAUTIFY_OPTIONS = {
  autosemicolon: true,
  indent: '  ',
};

/**
 * is the runtime in the production environment
 */
export const IS_PRODUCTION =
  typeof process !== 'undefined' && process.env?.NODE_ENV === 'production';
