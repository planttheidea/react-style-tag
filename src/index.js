// external dependencies
import React, {
  Component,
  PropTypes
} from 'react';

// local utils
import hashKeys from './hash';
import {
  minify,
  prefixCss
} from './transform';

const URL = window.URL || window.webkitURL;
const HAS_BLOB_SUPPORT = !!(window.Blob && typeof window.Blob === 'function' && URL.createObjectURL) && ((win) => {
  try {
    new win.Blob;

    return true;
  } catch (exception) {
    return false;
  }
})(window);
const NO_BLOB_SUPPORT_ERROR = (`
To support sourcemaps for react-style-tag you need Blob support, and the browser you are using does not currently 
support it. Please import the included polyfill at 'react-style-tag/blob-polyfill.js'.
`).replace(/\r?\n|\r/, '');
const ONLY_TEXT_ERROR = 'The only type of child that can be used in the <Style/> tag is text.';

/**
 * get the (if applicable) prefixed and minified css based on the
 * original cssText
 *
 * @param {string} cssText
 * @param {boolean} doNotPrefix
 * @param {boolean} isMinified
 * @returns {string}
 */
const getTransformedCss = (cssText, doNotPrefix, isMinified) => {
  const transformedCss = doNotPrefix ? cssText : prefixCss(cssText);

  return isMinified ? minify(transformedCss) : transformedCss;
};

/**
 * return the propsValue if it exists, else return the defaultValue
 * 
 * @param {boolean} propsValue
 * @param {boolean} defaultValue
 * @returns {boolean}
 */
const getCoalescedPropsValue = (propsValue, defaultValue) => {
  if (isUndefined(propsValue)) {
    return defaultValue;
  }
  
  return propsValue;
};

/**
 * determine if object is a boolean
 *
 * @param {*} object
 * @returns {boolean}
 */
const isBoolean = (object) => {
  return object === true || object === false;
};

/**
 * determine if object is a string
 *
 * @param {*} object
 * @returns {boolean}
 */
const isString = (object) => {
  return Object.prototype.toString.call(object) === '[object String]';
};

/**
 * determine if object is undefined
 *
 * @param {*} object
 * @returns {boolean}
 */
const isUndefined = (object) => {
  return object === void 0;
};

/**
 * throw an error if the provided children is not a text node
 *
 * @param {*} children
 */
const throwErrorIfIsNotText = (children) => {
  if (!isString(children)) {
    throw new Error(ONLY_TEXT_ERROR);
  }
};

const REACT_STYLE_TAG_GLOBAL_PROPERTIES = {
  doNotPrefix: false,
  hasSourceMap: false,
  isMinified: false
};

class Style extends Component {
  static propTypes = {
    children: PropTypes.node,
    doNotPrefix: PropTypes.bool,
    hasSourceMap: PropTypes.bool,
    isMinified: PropTypes.bool
  };

  componentDidMount() {
    this.setCorrectTag();
  }

  componentDidUpdate() {
    this.setCorrectTag();
  }

  componentWillUnmount() {
    this.removeTagFromHead('link');
    this.removeTagFromHead('style');
  }

  shouldComponentUpdate({children: nextChildren}) {
    const {
      children
    } = this.props;

    return children !== nextChildren;
  }

  /**
   * set the global options for all instances of Style
   *
   * @param {object} options
   * @param {boolean} [options.doNotPrefix]
   * @param {boolean} [options.hasSourceMap]
   * @param {boolean} [options.isMinified]
   */
  static setGlobalOptions(options) {
    Object.keys(options).forEach((option) => {
      if (REACT_STYLE_TAG_GLOBAL_PROPERTIES.hasOwnProperty(option)) {
        if (!isBoolean(options[option])) {
          throw new Error(`${option} must be a boolean value.`);
        }

        REACT_STYLE_TAG_GLOBAL_PROPERTIES[option] = options[option];
      }
    });
  }

  /**
   * remove the tagType from the document head if it exists
   *
   * @param {string} tagType
   */
  removeTagFromHead = (tagType) => {
    const tag = this.refs[tagType];

    if (tag) {
      document.head.removeChild(tag);
    }
  };

  /**
   * based on whether sourcemaps are requested, set either a link or style tag
   */
  setCorrectTag = () => {
    const {
      hasSourceMap
    } = this.props;
    
    const {
      hasSourceMap: hasSourceMapGlobal
    } = REACT_STYLE_TAG_GLOBAL_PROPERTIES;
    
    const hasSourceMapFinal = getCoalescedPropsValue(hasSourceMap, hasSourceMapGlobal);

    if (hasSourceMapFinal) {
      this.setLinkTag();
    } else {
      this.setStyleTag();
    }
  };

  /**
   * set the link tag with the prefixed / minified css text as a blob and move to the document head
   */
  setLinkTag = () => {
    const {
      children,
      doNotPrefix,
      isMinified
    } = this.props;

    throwErrorIfIsNotText(children);

    this.removeTagFromHead('style');

    if (HAS_BLOB_SUPPORT) {
      const {
        doNotPrefix: doNotPrefixGlobal,
        isMinified: isMinifiedGlobal
      } = REACT_STYLE_TAG_GLOBAL_PROPERTIES;

      const doNotPrefixFinal = getCoalescedPropsValue(doNotPrefix, doNotPrefixGlobal);
      const isMinifiedFinal = getCoalescedPropsValue(isMinified, isMinifiedGlobal);
      const link = this.refs.link;
      const transformedCss = getTransformedCss(children, doNotPrefixFinal, isMinifiedFinal);
      const blob = new window.Blob([transformedCss], {
        type: 'text/css'
      });

      link.href = URL.createObjectURL(blob);

      document.head.appendChild(link);
    } else {
      throw new Error(NO_BLOB_SUPPORT_ERROR);
    }
  };

  /**
   * set the style tag with the prefixed / minified css text and move to the document head
   */
  setStyleTag = () => {
    const {
      children,
      doNotPrefix,
      isMinified
    } = this.props;

    const {
      doNotPrefix: doNotPrefixGlobal,
      isMinified: isMinifiedGlobal
    } = REACT_STYLE_TAG_GLOBAL_PROPERTIES;

    throwErrorIfIsNotText(children);

    this.removeTagFromHead('link');

    const doNotPrefixFinal = getCoalescedPropsValue(doNotPrefix, doNotPrefixGlobal);
    const isMinifiedFinal = getCoalescedPropsValue(isMinified, isMinifiedGlobal);
    const style = this.refs.style;

    style.textContent = getTransformedCss(children, doNotPrefixFinal, isMinifiedFinal);

    document.head.appendChild(style);
  };

  render() {
    const {
      children,
      doNotPrefix: doNotPrefixIgnored,
      hasSourceMap,
      isMinified: isMinifiedIgnored,
      ...otherProps
    } = this.props;

    const {
      hasSourceMap: hasSourceMapGlobal
    } = REACT_STYLE_TAG_GLOBAL_PROPERTIES;

    const hasSourceMapFinal = getCoalescedPropsValue(hasSourceMap, hasSourceMapGlobal);

    if (hasSourceMapFinal && HAS_BLOB_SUPPORT) {
      return (
        <link
          rel="stylesheet"
          ref="link"
          {...otherProps}
        />
      );
    }

    return (
      <style
        ref="style"
        {...otherProps}
      >
        {children}
      </style>
    );
  }
}

export {hashKeys};

export default Style;
