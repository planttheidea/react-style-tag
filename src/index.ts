export { Style } from './Style';
export { hashKeys } from './hash';
export { getGlobalOptions, setGlobalOptions } from './options';

export type { Options, Props as StyleProps } from '../index.d';
