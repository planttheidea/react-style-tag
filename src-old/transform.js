// external dependencies
import compose from 'lodash/fp/compose';
import isUndefined from 'lodash/isUndefined';
import autoprefixer from 'autoprefixer';
import postcss from 'postcss';

// constants
import {DEFAULT_AUTOPREFIXER_OPTIONS} from './constants';

/**
 * @function getAutoprefixer
 *
 * @description
 * Function that returns an autoprefixer instance configured with the provided options.
 * Memoized for better performance
 *
 * @param {object} options
 * @returns {object}
 */
export const getAutoprefixer = (() => {
  const cache = {
    options: null,
    postcss: null
  };

  return (autoprefixerOptions) => {
    if (cache.postcss && cache.options === autoprefixerOptions) {
      return cache.postcss;
    }

    cache.options = autoprefixerOptions;

    return (cache.postcss = postcss([autoprefixer(autoprefixerOptions)]));
  };
})();

/**
 * @function getCoalescedPropsValue
 *
 * @description
 * return the propsValue if it exists, else return the defaultValue
 *
 * @param {boolean} propsValue
 * @param {boolean} defaultValue
 * @returns {boolean}
 */
export const getCoalescedPropsValue = (propsValue, defaultValue) => {
  return isUndefined(propsValue) ? defaultValue : propsValue;
};

/**
 * @function minify
 *
 * @description
 * return the minified string css
 *
 * @param {string} cssText
 * @returns {string}
 */
export const minify = (cssText) => {
  return cssText
    .trim()
    .replace(/\/\*[\s\S]+?\*\//g, '')
    .replace(/[\n\r]/g, '')
    .replace(/\s*([:;,{}])\s*/g, '$1')
    .replace(/\s+/g, ' ')
    .replace(/;}/g, '}')
    .replace(/\s+(!important)/g, '$1')
    .replace(/#([a-fA-F0-9])\1([a-fA-F0-9])\2([a-fA-F0-9])\3(?![a-fA-F0-9])/g, '#$1$2$3')
    .replace(/(Microsoft[^;}]*)#([a-fA-F0-9])([a-fA-F0-9])([a-fA-F0-9])(?![a-fA-F0-9])/g, '$1#$2$2$3$3$4$4')
    .replace(/\b(\d+[a-z]{2}) \1 \1 \1/gi, '$1')
    .replace(/\b(\d+[a-z]{2}) (\d+[a-z]{2}) \1 \2/gi, '$1 $2')
    .replace(/([\s|:])[0]+px/g, '$10');
};

/**
 * @function prefixCss
 *
 * @description
 * return the css after running through autoprefixer
 *
 * @param {string} cssText
 * @param {object} autoprefixerOptions
 * @returns {string}
 */
export const prefixCss = (cssText, autoprefixerOptions) => {
  return getAutoprefixer(autoprefixerOptions).process(cssText).css;
};

/**
 * @function prefixAndMinifyCss
 *
 * @description
 * return the css after running through autoprefixer and minify
 *
 * @param {string} cssText
 * @returns {string}
 */
export const prefixAndMinifyCss = compose(minify, prefixCss);

/**
 * @function getTransformedCss
 *
 * @description
 * get the (if applicable) prefixed and minified css based on the
 * original cssText
 *
 * @param {string} cssText
 * @param {boolean} doNotPrefix=false
 * @param {boolean} isMinified=false
 * @param {object} autoprefixerOptions=DEFAULT_AUTOPREFIXER_OPTIONS
 * @returns {string}
 */
export const getTransformedCss = (
  cssText,
  doNotPrefix = false,
  isMinified = false,
  autoprefixerOptions = DEFAULT_AUTOPREFIXER_OPTIONS
) => {
  if (!isMinified) {
    return doNotPrefix ? cssText : prefixCss(cssText, autoprefixerOptions);
  }

  return doNotPrefix ? minify(cssText) : prefixAndMinifyCss(cssText, autoprefixerOptions);
};
