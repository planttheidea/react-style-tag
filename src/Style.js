// external dependencies
import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import {createElementRef, createMethod} from 'react-parm';

// blob
import {createGetCachedLinkHref, hasBlobSupport} from './blob';

// constants
import {SUPPORTS_BEFORE_UPDATE_SNAPSHOT} from './constants';

// options
import {getCoalescedOption, setGlobalOptions} from './options';

// styles
import {getRenderedStyles} from './styles';

/**
 * @function componentDidMount
 *
 * @description
 * on mount, relocate the node
 *
 * @param {ReactComponent} instance the component instance
 * @param {HTMLElement} instance.node the node to render the styles into
 * @param {function} instance.relocateNote the method to relocate the node to the head
 * @returns {void}
 */
export const componentDidMount = ({node, relocateNode}) => relocateNode(node);

/**
 * @function getSnapshotBeforeUpdate
 *
 * @description
 * before the update occurs, if the sourceMap requirements have changed, return the node to its original position
 *
 * @param {ReactComponent} instance the component instance
 * @param {HTMLElement} instance.node the node to render the styles into
 * @param {function} instance.returnNode the method to return the node to its original parent
 * @returns {null}
 */
export const getSnapshotBeforeUpdate = ({node, returnNode}) => {
  returnNode(node);

  return null;
};

/**
 * @function componentDidUpdate
 *
 * @description
 * on update, if the sourceMap requirements have changed then relocate the new node to the head,
 * and if the styles have changed then set them in staet
 *
 * @param {ReactComponent} instance the component instance
 * @param {function} instance.getStyleForState the method to get the new rendered style
 * @param {HTMLElement} instance.node the node to render the styles into
 * @param {function} instance.relocateNote the method to relocate the node to the head
 * @param {Object} instance.props the new props of the component
 * @param {function} instance.setState the method to set the state of the component
 * @param {Array<any>} args the arguments passed to the function
 * @param {Object} previousProps the previous props of the component
 */
export const componentDidUpdate = ({getStyleForState, node, relocateNode, props, setState}, [previousProps]) => {
  relocateNode(node);

  if (props.children !== previousProps.children) {
    setState(getStyleForState);
  }
};

/**
 * @function componentWillUnmount
 *
 * @description
 * prior to unmount, return the node to its original parent
 *
 * @param {ReactComponent} instance the component instance
 * @param {HTMLElement} instance.node the node to render the styles into
 * @param {function} instance.returnNode the method to return the node to its original parent
 * @returns {void}
 */
export const componentWillUnmount = ({node, returnNode}) => returnNode(node);

/**
 * @function getStyleForState
 *
 * @description
 * get the styles to be used for the rendered tag
 *
 * @param {ReactComponent} instance the component instance
 * @param {Object} instance.props the new props of the component
 * @returns {{style: string}} the style strng to use in the rendered tag
 */
export const getStyleForState = ({props}) => ({
  style: getRenderedStyles(props.children, {
    isCompressed: getCoalescedOption(props, 'isCompressed'),
    isMinified: getCoalescedOption(props, 'isMinified'),
    isPrefixed: getCoalescedOption(props, 'isPrefixed')
  })
});

/**
 * @function relocateNode
 *
 * @description
 * relocate the node to the bottom of the head
 *
 * @param {ReactComponent} instance the component instance
 * @param {Array<any>} args the arguments passed to the function
 * @param {HTMLElement} node the node to render the styles into
 */
export const relocateNode = (instance, [node]) => {
  if (typeof document !== 'undefined' && node) {
    instance.originalParent = node.parentNode;

    instance.originalParent.removeChild(node);
    document.head.appendChild(node);
  }
};

/**
 * @function returnNode
 *
 * @description
 * return the node to the original parent
 *
 * @param {ReactComponent} instance the component instance
 * @param {Array<any>} args the arguments passed to the function
 * @param {HTMLElement} node the node to render the styles into
 */
export const returnNode = (instance, [node]) => {
  if (typeof document !== 'undefined' && node) {
    try {
      document.head.removeChild(node);
      instance.originalParent.appendChild(node);
    } catch (error) {
      // ignore the error
    } finally {
      instance.node = null;
      instance.originalParent = null;
    }
  }
};

class Style extends PureComponent {
  static propTypes = {
    children: PropTypes.string.isRequired,
    hasSourceMap: PropTypes.bool,
    id: PropTypes.string,
    isCompressed: PropTypes.bool,
    isMinified: PropTypes.bool,
    isPrefixed: PropTypes.bool
  };

  constructor(props) {
    super(props);

    this.state = getStyleForState({props});
  }

  // lifecycle methods
  componentDidMount = createMethod(this, componentDidMount);
  componentDidUpdate = createMethod(this, componentDidUpdate);
  /* eslint-disable react/sort-comp */
  [SUPPORTS_BEFORE_UPDATE_SNAPSHOT ? 'getSnapshotBeforeUpdate' : 'componentWillUpdate'] = createMethod(
    this,
    getSnapshotBeforeUpdate
  );
  /* eslint-enable */
  componentWillUnmount = createMethod(this, componentWillUnmount);

  // instance values
  linkHref = null;
  node = null;
  originalParent = null;

  // static methods
  static setGlobalOptions = setGlobalOptions;

  // instance methods
  getCachedLinkHref = createGetCachedLinkHref();
  getStyleForState = createMethod(this, getStyleForState);
  relocateNode = createMethod(this, relocateNode);
  returnNode = createMethod(this, returnNode);

  render() {
    const {
      children: childrenIgnored,
      hasSourceMap: hasSourceMapIgnored,
      isCompressed: isMinifiedIgnored,
      isPrefixed: isPrefixedIgnored,
      ...props
    } = this.props;
    const {style} = this.state;

    if (getCoalescedOption(this.props, 'hasSourceMap')) {
      if (hasBlobSupport()) {
        return (
          /* eslint-disable prettier */
          <link
            {...props}
            href={this.getCachedLinkHref(style)}
            ref={createElementRef(this, 'node')}
            rel="stylesheet"
          />
          /* eslint-enable */
        );
      }

      /* eslint-disable no-console */
      console.error(
        'To support sourcemaps for react-style-tag you need Blob support, and the browser you are using does not currently support it. You should include a polyfill prior to the rendering of this component.'
      );
      /* eslint-enable */
    }

    return (
      <style
        ref={createElementRef(this, 'node')}
        {...props}
      >
        {style}
      </style>
    );
  }
}

export default Style;
