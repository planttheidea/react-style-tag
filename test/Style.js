// test
import test from 'ava';
import _ from 'lodash';
import React from 'react';
import sinon from 'sinon';
import {
  shallow
} from 'enzyme';
import toJson from 'enzyme-to-json';

// src
import * as component from 'src/Style';
import * as cache from 'src/cache';
import * as constants from 'src/constants';
import * as transform from 'src/transform';
import * as utils from 'src/utils';

const LINK_HREF = '__LINK_HREF__';

test('if createAssignRefToInstance will assign a ref to the instance at refName', (t) => {
  const refName = 'foo';
  const instance = {
    [refName]: null
  };

  const assignRefToInstance = component.createAssignRefToInstance(instance, refName);

  t.true(_.isFunction(assignRefToInstance));

  const element = {};

  assignRefToInstance(element);

  t.is(instance[refName], element);
  t.deepEqual(instance, {
    [refName]: element
  });
});

test('if createComponentDidMount will call setCorrectTag on the instance', (t) => {
  const instance = {
    setCorrectTag: sinon.spy()
  };

  const componentDidMount = component.createComponentDidMount(instance);

  t.true(_.isFunction(componentDidMount));

  componentDidMount();

  t.true(instance.setCorrectTag.calledOnce);
});

test('if createComponentDidUpdate calls setCacheId and setCorrectTag', (t) => {
  const instance = {
    id: 'foo',
    props: {
      children: 'bar'
    },
    setCorrectTag: sinon.spy()
  };

  const componentDidUpdate = component.createComponentDidUpdate(instance);

  t.true(_.isFunction(componentDidUpdate));

  const stub = sinon.stub(cache, 'setCacheId');

  componentDidUpdate();

  t.true(stub.calledOnce);
  t.true(stub.calledWith(instance.id, instance.props.children));

  t.true(instance.setCorrectTag.calledOnce);

  stub.restore();
});

test.serial('if createComponentWillMount will only call getUrl and createIdForTag if blob support is not present', (t) => {
  const instance = {
    id: null,
    props: {
      children: 'bar',
      id: 'foo'
    }
  };

  const componentWillMount = component.createComponentWillMount(instance);

  t.true(_.isFunction(componentWillMount));

  const urlStub = sinon.stub(utils, 'getUrl');
  const blobSupportStub = sinon.stub(utils, 'getHasBlobSupport');

  const createIdStub = sinon.stub(cache, 'createIdForTag').returns(instance.props.id);

  componentWillMount();

  t.true(urlStub.calledOnce);

  t.true(blobSupportStub.notCalled);

  t.true(createIdStub.calledOnce);
  t.true(createIdStub.calledWith(instance.props.id, instance.props.children));

  urlStub.restore();
  blobSupportStub.restore();
  createIdStub.restore();
});

test.serial('if createComponentWillMount will call get getHasBlobSupport if url exists', (t) => {
  const instance = {
    id: null,
    props: {
      children: 'bar',
      id: 'foo'
    }
  };

  const componentWillMount = component.createComponentWillMount(instance);

  t.true(_.isFunction(componentWillMount));

  const urlStub = sinon.stub(utils, 'getUrl').returns({
    createObjectURL: sinon.stub().returns(LINK_HREF)
  });
  const blobSupportStub = sinon.stub(utils, 'getHasBlobSupport').returns(true);

  const createIdStub = sinon.stub(cache, 'createIdForTag').returns(instance.props.id);

  componentWillMount();

  t.true(urlStub.calledOnce);

  t.true(blobSupportStub.calledOnce);

  t.true(createIdStub.calledOnce);
  t.true(createIdStub.calledWith(instance.props.id, instance.props.children));

  urlStub.restore();
  blobSupportStub.restore();
  createIdStub.restore();
});

test('if createComponentWillUnmount will call removeTagFromHead and removeIdFromCache', (t) => {
  const instance = {
    id: 'foo',
    removeTagFromHead: sinon.spy()
  };

  const componentWillUnmount = component.createComponentWillUnmount(instance);

  t.true(_.isFunction(componentWillUnmount));

  const removeIdStub = sinon.stub(cache, 'removeIdFromCache');

  componentWillUnmount();

  t.true(instance.removeTagFromHead.calledTwice);
  t.is(instance.removeTagFromHead.getCall(0).args[0], 'link');
  t.is(instance.removeTagFromHead.getCall(1).args[0], 'style');

  t.true(removeIdStub.calledOnce);
  t.true(removeIdStub.calledWith(instance.id));

  removeIdStub.restore();
});

test('if createComponentWillUnmount will mpt call removeIdFromCache if id is null', (t) => {
  const instance = {
    id: null,
    removeTagFromHead: sinon.spy()
  };

  const componentWillUnmount = component.createComponentWillUnmount(instance);

  t.true(_.isFunction(componentWillUnmount));

  const removeIdStub = sinon.stub(cache, 'removeIdFromCache');

  componentWillUnmount();

  t.true(instance.removeTagFromHead.calledTwice);
  t.is(instance.removeTagFromHead.getCall(0).args[0], 'link');
  t.is(instance.removeTagFromHead.getCall(1).args[0], 'style');

  t.true(removeIdStub.notCalled);

  removeIdStub.restore();
});

test.serial('if createRemoveTagFromHead will create a method that that gets the tag from the instance and removes it from the head', (t) => {
  const tagType = 'foo';
  const tag = {};
  const instance = {
    [tagType]: tag
  };

  const removeTagFromHead = component.createRemoveTagFromHead(instance);

  t.true(_.isFunction(removeTagFromHead));

  const removeChildStub = sinon.stub(document.head, 'removeChild');

  removeTagFromHead(tagType);

  t.true(removeChildStub.calledOnce);
  t.true(removeChildStub.calledWith(tag));
  t.is(instance[tagType], null);

  removeChildStub.restore();
});

test('if createSetCorrectTag will call the right tag function based on hasSourceMap being false', (t) => {
  const instance = {
    getCoalescedPropsValue: sinon.stub().returns(false),
    id: 'foo',
    props: {
      hasSourceMap: undefined
    },
    setLinkTag: sinon.spy(),
    setStyleTag: sinon.spy()
  };

  const setCorrectTag = component.createSetCorrectTag(instance);

  t.true(_.isFunction(setCorrectTag));

  setCorrectTag();

  t.true(instance.getCoalescedPropsValue.calledOnce);

  t.true(instance.setLinkTag.notCalled);

  t.true(instance.setStyleTag.calledOnce);
});

test('if createSetCorrectTag will call the right tag function based on hasSourceMap being true', (t) => {
  const instance = {
    getCoalescedPropsValue: sinon.stub().returns(true),
    id: 'foo',
    props: {
      hasSourceMap: undefined
    },
    setLinkTag: sinon.spy(),
    setStyleTag: sinon.spy()
  };

  const setCorrectTag = component.createSetCorrectTag(instance);

  t.true(_.isFunction(setCorrectTag));

  setCorrectTag();

  t.true(instance.getCoalescedPropsValue.calledOnce);

  t.true(instance.setLinkTag.calledOnce);

  t.true(instance.setStyleTag.notCalled);
});

test('if createSetCorrectTag will call nothing if the instance id is null', (t) => {
  const instance = {
    getCoalescedPropsValue: sinon.stub().returns(true),
    id: null,
    props: {
      hasSourceMap: undefined
    },
    setLinkTag: sinon.spy(),
    setStyleTag: sinon.spy()
  };

  const setCorrectTag = component.createSetCorrectTag(instance);

  t.true(_.isFunction(setCorrectTag));

  setCorrectTag();

  t.true(instance.getCoalescedPropsValue.notCalled);

  t.true(instance.setLinkTag.notCalled);

  t.true(instance.setStyleTag.notCalled);
});

test.serial('if createSetLinkTag will create a link tag with the correct values', (t) => {
  const instance = {
    getCoalescedPropsValue: sinon.stub().returns(false),
    link: {
      href: null
    },
    props: {
      children: 'foo',
      doNotPrefix: false,
      isMinified: false
    },
    removeTagFromHead: sinon.spy()
  };

  const setLinkTag = component.createSetLinkTag(instance);

  t.true(_.isFunction(setLinkTag));

  const transformCssStub = sinon.stub(transform, 'getTransformedCss').returns(instance.props.children);

  global.window = {
    Blob: sinon.spy()
  };

  const appendChildStub = sinon.stub(document.head, 'appendChild');

  component.createComponentWillMount(instance)();

  t.notThrows(() => {
    setLinkTag();
  });

  transformCssStub.restore();

  t.true(instance.removeTagFromHead.calledOnce);
  t.true(instance.removeTagFromHead.calledWith('style'));

  t.true(window.Blob.calledOnce);
  t.deepEqual(window.Blob.getCall(0).args, [
    [instance.props.children],
    {
      type: 'text/css'
    }
  ]);

  t.is(instance.link.href, LINK_HREF);

  t.true(appendChildStub.calledOnce);
  t.true(appendChildStub.calledWith(instance.link));

  global.window.Blob = undefined;

  appendChildStub.restore();
});

test.serial('if createSetStyleTag will create a link tag with the correct values', (t) => {
  const instance = {
    getCoalescedPropsValue: sinon.stub().returns(false),
    style: {
      textContent: null
    },
    props: {
      children: 'foo',
      doNotPrefix: false,
      isMinified: false
    },
    removeTagFromHead: sinon.spy()
  };

  const setStyleTag = component.createSetStyleTag(instance);

  t.true(_.isFunction(setStyleTag));

  const transformCssStub = sinon.stub(transform, 'getTransformedCss').returns(instance.props.children);
  const appendChildStub = sinon.stub(document.head, 'appendChild');

  component.createComponentWillMount(instance)();

  t.notThrows(() => {
    setStyleTag();
  });

  transformCssStub.restore();

  t.true(instance.removeTagFromHead.calledOnce);
  t.true(instance.removeTagFromHead.calledWith('link'));

  t.true(transformCssStub.calledOnce);

  t.is(instance.style.textContent, instance.props.children);

  t.true(appendChildStub.calledOnce);
  t.true(appendChildStub.calledWith(instance.style));

  appendChildStub.restore();
});

test('if createShouldComponentUpdate correctly compares children that are the same', (t) => {
  const instance = {
    props: {
      children: 'foo'
    }
  };

  const shouldComponentUpdate = component.createShouldComponentUpdate(instance);

  t.true(_.isFunction(shouldComponentUpdate));

  const nextProps = {
    children: 'foo'
  };

  const result = shouldComponentUpdate(nextProps);

  t.false(result);
});

test('if createShouldComponentUpdate correctly compares children that are different', (t) => {
  const instance = {
    props: {
      children: 'foo'
    }
  };

  const shouldComponentUpdate = component.createShouldComponentUpdate(instance);

  t.true(_.isFunction(shouldComponentUpdate));

  const nextProps = {
    children: 'bar'
  };

  const result = shouldComponentUpdate(nextProps);

  t.true(result);
});

test('if setGlobalOptions have the correct defaults', (t) => {
  const result = component.setGlobalOptions({});

  t.deepEqual(result, constants.DEFAULT_REACT_STYLE_TAG_GLOBAL_PROPERTIES);
});

test('if setGlobalOptions will set the global properties based on the options passed', (t) => {
  const overrides = {
    doNotPrefix: true,
    isMinified: true
  };

  const result = component.setGlobalOptions(overrides);

  t.deepEqual(result, {
    ...constants.DEFAULT_REACT_STYLE_TAG_GLOBAL_PROPERTIES,
    doNotPrefix: true,
    isMinified: true
  });
});

test('if setGlobalOptions will throw a TypeError when an override is not a boolean', (t) => {
  const overrides = {
    doNotPrefix: 'foo'
  };

  t.throws(() => {
    component.setGlobalOptions(overrides);
  }, TypeError);
});

test('if Style will render correctly when using a style tag', (t) => {
  const Style = component.default;
  const id = 'foo';

  const wrapper = shallow(
    <Style
      id={id}
    />
  );

  t.snapshot(toJson(wrapper));
});

test('if Style will render correctly when using a link tag', (t) => {
  const Style = component.default;
  const id = 'foo';

  const wrapper = shallow(
    <Style
      hasSourceMap
      id={id}
    />
  );

  t.snapshot(toJson(wrapper));
});

test('if Style will render correctly when id is null', (t) => {
  const Style = component.default;
  const id = null;

  const wrapper = shallow(
    <Style
      hasSourceMap
      id={id}
    />
  );

  t.is(wrapper.type(), null);
});
