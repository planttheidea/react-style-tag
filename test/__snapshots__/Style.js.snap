exports[`if Style will render correctly when using a link tag 1`] = `"{"__ava_react_jsx":{"type":"link","props":{"id":"foo","rel":"stylesheet"},"children":null}}"`;

exports[`if Style will render correctly when using a style tag 1`] = `"{\"__ava_react_jsx\":{\"type\":\"style\",\"props\":{\"id\":\"foo\"},\"children\":null}}"`;
